cgi.go

```diff
@@ -33,22 +36,44 @@ const (
 	sslProtocol
 )
 
-func allocServerVariable(cArr *[27]*C.char, env map[string]string, serverKey serverKey, envKey string, val string) {
-	if val, ok := env[envKey]; ok {
-		cArr[serverKey] = C.CString(val)
-		delete(env, envKey)
+var knownServerKeys = map[string]struct{}{
+	"CONTENT_LENGTH\x00":    {},
+	"DOCUMENT_ROOT\x00":     {},
+	"DOCUMENT_URI\x00":      {},
+	"GATEWAY_INTERFACE\x00": {},
+	"HTTP_HOST\x00":         {},
+	"HTTPS\x00":             {},
+	"PATH_INFO\x00":         {},
+	"PHP_SELF\x00":          {},
+	"REMOTE_ADDR\x00":       {},
+	"REMOTE_HOST\x00":       {},
+	"REMOTE_PORT\x00":       {},
+	"REQUEST_SCHEME\x00":    {},
+	"SCRIPT_FILENAME\x00":   {},
+	"SCRIPT_NAME\x00":       {},
+	"SERVER_NAME\x00":       {},
+	"SERVER_PORT\x00":       {},
+	"SERVER_PROTOCOL\x00":   {},
+	"SERVER_SOFTWARE\x00":   {},
+	"SSL_PROTOCOL\x00":      {},
+}
 
+func setKnownServerVariable(p *runtime.Pinner, cArr *[27]C.go_string, serverKey serverKey, val string) {
+	if val == "" {
 		return
 	}
 
-	cArr[serverKey] = C.CString(val)
+	valData := unsafe.StringData(val)
+	p.Pin(valData)
```

Comment:
Does this work? It seems like we would need to pin the Go structure here (`val`), and not the underlying array.

Reply:
Yes this works, but IIRC it's not necessary anymore in Go 1.22, we should now be able to pin the string directly.

Reply:
Oh nice! I should read the release notes :)

Reply:
Actually, this hack is still necessary: https://github.com/golang/go/issues/65286